File: sideai/backend/services/app_context.py
"""Detect what kind of app/page the user is on (email vs coding vs browser)."""

from __future__ import annotations

import re

# ── Strong signals ─────────────────────────────────────────────────────────────

_EMAIL_URL_RE = re.compile(
    r"mail\.google\.com|"
    r"outlook\.(live|office)\.com|"
    r"mail\.yahoo\.com|"
    r"mail\.proton\.me|"
    r"mail\.aol\.com|"
    r"webmail\.|"
    r"/mail/",
    re.I,
)

_NON_EMAIL_URL_RE = re.compile(
    r"neetcode\.io|leetcode\.com|hackerrank\.com|codewars\.com|"
    r"github\.com|gitlab\.com|stackoverflow\.com|"
    r"notion\.so|figma\.com|docs\.google\.com|"
    r"youtube\.com|youtu\.be|"
    r"neetcode\.io|geeksforgeeks\.org|codeforces\.com",
    re.I,
)

_GMAIL_TITLE_RE = re.compile(
    r"\bgmail\b|mail\.google\.com|inbox\s*\(|inbox\s*-",
    re.I,
)

_OUTLOOK_TITLE_RE = re.compile(
    r"\boutlook\b.*(@|\||-)|microsoft outlook",
    re.I,
)

_MAIL_APP_RE = re.compile(
    r"^(mail|apple mail|thunderbird|superhuman|microsoft outlook)$",
    re.I,
)

_CODING_IN_TEXT_RE = re.compile(
    r"neetcode\.io|leetcode\.com|hackerrank|codewars|"
    r"products of array|submissions|discuss\s+question|"
    r"run\s+code|submit\s+solution|time complexity|space complexity",
    re.I,
)


def _norm(s: str) -> str:
    return (s or "").strip()


def is_coding_context(
    active_app: str = "",
    window_title: str = "",
    url: str | None = None,
    visible_text: str = "",
) -> bool:
    blob = f"{_norm(active_app)} {_norm(window_title)} {url or ''} {_norm(visible_text)[:600]}".lower()
    if _NON_EMAIL_URL_RE.search(blob) and not _EMAIL_URL_RE.search(blob):
        return True
    if _CODING_IN_TEXT_RE.search(blob):
        return True
    title = _norm(window_title).lower()
    if "neetcode" in title or "leetcode" in title:
        return True
    return False


def is_email_context(
    active_app: str = "",
    window_title: str = "",
    url: str | None = None,
    visible_text: str = "",
) -> bool:
    """
    True only when the user is in a mail client or webmail inbox — not coding sites,
    and not generic pages that mention the word 'email'.
    """
    active_app = _norm(active_app)
    window_title = _norm(window_title)
    url = _norm(url or "")
    snippet = _norm(visible_text)[:800]

    if is_coding_context(active_app, window_title, url, snippet):
        return False

    if url:
        if _EMAIL_URL_RE.search(url):
            return True
        if _NON_EMAIL_URL_RE.search(url):
            return False

    app_key = active_app.lower()
    if _MAIL_APP_RE.match(app_key) or app_key in ("mail", "thunderbird", "superhuman"):
        return True
    if "outlook" in app_key and "visual studio" not in app_key:
        return True

    if _GMAIL_TITLE_RE.search(window_title):
        return True
    if _OUTLOOK_TITLE_RE.search(window_title):
        return True

    # Webmail hints in captured text (Gmail UI), not bare "email" substring
    if snippet:
        low = snippet.lower()
        if "mail.google.com" in low or "inbox" in low and "compose" in low and "gmail" in low:
            return True

    return False


def is_browser_context(active_app: str = "", window_title: str = "") -> bool:
    blob = f"{_norm(active_app)} {_norm(window_title)}".lower()
    browsers = (
        "chrome", "safari", "firefox", "arc", "brave", "edge", "opera", "vivaldi", "chromium",
    )
    return any(b in blob for b in browsers)

File: sideai/frontend/src/utils/appContext.ts
/** Detect email vs coding vs browser context — keep in sync with backend app_context.py */

const EMAIL_URL_RE =
  /mail\.google\.com|outlook\.(live|office)\.com|mail\.yahoo\.com|mail\.proton\.me|webmail\./i;

const NON_EMAIL_URL_RE =
  /neetcode\.io|leetcode\.com|hackerrank\.com|codewars\.com|github\.com|gitlab\.com|stackoverflow\.com|youtube\.com|youtu\.be/i;

const GMAIL_TITLE_RE = /\bgmail\b|mail\.google\.com|inbox\s*[\(-]/i;

const OUTLOOK_TITLE_RE = /\boutlook\b.*(@|\||-)|microsoft outlook/i;

const MAIL_APP_RE = /^(mail|apple mail|thunderbird|superhuman|microsoft outlook)$/i;

const CODING_TEXT_RE =
  /neetcode\.io|leetcode\.com|hackerrank|codewars|products of array|submissions|discuss\s+question|run\s+code|submit\s+solution/i;

function norm(s: string) {
  return (s || "").trim();
}

export function isCodingContext(
  activeApp = "",
  windowTitle = "",
  url = "",
  visibleText = ""
): boolean {
  const blob = `${norm(activeApp)} ${norm(windowTitle)} ${url} ${norm(visibleText).slice(0, 600)}`.toLowerCase();
  if (NON_EMAIL_URL_RE.test(blob) && !EMAIL_URL_RE.test(blob)) return true;
  if (CODING_TEXT_RE.test(blob)) return true;
  const title = norm(windowTitle).toLowerCase();
  if (title.includes("neetcode") || title.includes("leetcode")) return true;
  return false;
}

export function isEmailContext(
  activeApp = "",
  windowTitle = "",
  url = "",
  visibleText = ""
): boolean {
  const app = norm(activeApp);
  const title = norm(windowTitle);
  const u = norm(url);
  const snippet = norm(visibleText).slice(0, 800);

  if (isCodingContext(app, title, u, snippet)) return false;

  if (u) {
    if (EMAIL_URL_RE.test(u)) return true;
    if (NON_EMAIL_URL_RE.test(u)) return false;
  }

  const appKey = app.toLowerCase();
  if (MAIL_APP_RE.test(appKey) || appKey === "mail" || appKey === "thunderbird" || appKey === "superhuman") {
    return true;
  }
  if (appKey.includes("outlook") && !appKey.includes("visual studio")) return true;

  if (GMAIL_TITLE_RE.test(title)) return true;
  if (OUTLOOK_TITLE_RE.test(title)) return true;

  if (snippet) {
    const low = snippet.toLowerCase();
    if (low.includes("mail.google.com")) return true;
    if (low.includes("inbox") && low.includes("compose") && low.includes("gmail")) return true;
  }

  return false;
}

export function isBrowserContext(activeApp = "", windowTitle = ""): boolean {
  const blob = `${norm(activeApp)} ${norm(windowTitle)}`.toLowerCase();
  return ["chrome", "safari", "firefox", "arc", "brave", "edge", "opera", "vivaldi", "chromium"].some(
    (b) => blob.includes(b)
  );
}
